### DISCUSSION TITLE
Is it possible to apply a median filter to a mesh?

### DISCUSSION THREAD

**Discussion by koenterheegde0507** [pyvista/pyvista#5947] [Q&A]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
mesh = pv.PolyData(mesh_verts,faces_with_counts)
largest = mesh.connectivity('largest') 
grid = largest.cast_to_unstructured_grid()
...
new_mesh = grid.extract_surface()
```

### Description

I'm creating a PolyData based on vertices and faces that I get from the Marching Cubes algorithm.  I then extract the largest connected parts from it and plot it. I then use some callback function to edit parts of the mesh, that's why I have to cast it to an unstructured grid. After editing, I generate a new mesh using `extract_surface()`. I have been using Slicer3D in the past and you can smooth the mesh using a median filter. I have found that pyvista also has this functionality: (https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.imagedatafilters.median_smooth#pyvista.ImageDataFilters.median_smooth) but it's made for `ImageData` and not for PolyData or an Unstructured Grid. How could I implement this for my situation? 

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Fri Apr 19 12:11:58 2024 West-Europa (zomertijd)

                OS : Windows
            CPU(s) : 8
           Machine : AMD64
      Architecture : 64bit
               RAM : 7.8 GiB
       Environment : Python
        GPU Vendor : Intel
      GPU Renderer : Intel(R) UHD Graphics 620
       GPU Version : 4.5.0 - Build 30.0.101.1692
  MathText Support : False

  Python 3.9.7 (tags/v3.9.7:1016ef3, Aug 30 2021, 20:19:38) [MSC v.1929 64 bit
  (AMD64)]

           pyvista : 0.42.3
               vtk : 9.3.0
             numpy : 1.24.3
        matplotlib : 3.8.2
            scooby : 0.9.2
             pooch : 1.8.0
            pillow : 10.1.0
           imageio : 2.33.0
             PyQt5 : 5.15.4
           IPython : 8.13.2
        ipywidgets : 8.1.2
             scipy : 1.9.1
              tqdm : 4.66.2
            meshio : 5.3.5
      nest_asyncio : 1.5.6
--------------------------------------------------------------------------------
```

### Screenshots

![Screenshot 2024-04-19 121505](https://github.com/pyvista/pyvista/assets/152174547/ee1edfc1-8f41-432a-940b-11910b5fdde3)
Unsmoothed mesh that I get.

<img width="960" alt="Schermafbeelding 2024-03-13 123413" src="https://github.com/pyvista/pyvista/assets/152174547/19f61251-8d2b-43d7-8ccb-21a53c900cc3">
Mesh that I would like to get (Made in 3DSlicer)

**Comment by user27182** [comment]
> I'm creating a PolyData based on vertices and faces that I get from the Marching Cubes algorithm.

Marching Cubes is a fairly dated method for generating a surface mesh from image data.  If your application permits, you can try using a more modern surface generation method such as Surface Nets instead. PyVista has an implementation here: https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.ImageDataFilters.contour_labeled.html#pyvista.ImageDataFilters.contour_labeled

The nice thing about that method is it has smoothing built-in. So, depending on the kinds of edits you're making, this may work for you to both generate the surface and do the smoothing.

> I have been using Slicer3D in the past and you can smooth the mesh using a median filter. I have found that pyvista also has this functionality: (https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.imagedatafilters.median_smooth#pyvista.ImageDataFilters.median_smooth) but it's made for `ImageData` and not for PolyData or an Unstructured Grid. How could I implement this for my situation?

The median filter, both in PyVista and in 3DSlicer, is fundamentally an image filter. The reason you can use it to smooth a surface mesh with 3DSlicer is because 3DSlicer implements a synchronisation method that keeps all representations of your data in sync. If you edit the image data, the surface gets updated accordingly to reflect the changes. So in that way, you can use a median filter on the image data in 3DSlicer, and in the background 3DSlicer will filter the image, but then also generate a new surface for you automatically based on the new filtered image (it uses vtkFlyingEdges to generate the surface and applies smoothing by default). 

Unfortunately PyVista does not have any such mechanism to keep image data and surface data synchronized like this.

**Reply by user27182** [reply]
As a follow up, although you can't use a median filter on a surface mesh directly, you _can_ try the following workflow in PyVista.
1. start with noisy binary image data
2. use the [median smooth filter](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.imagedatafilters.median_smooth#pyvista.ImageDataFilters.median_smooth) on the image data to denoise it
3. optionally you can also try using dilation and erosion to further denoise or "smooth" the image data, see [image_dilate_erode](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.ImageDataFilters.image_dilate_erode.html#pyvista.ImageDataFilters.image_dilate_erode)
4. use the [contour_labeled filter](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.ImageDataFilters.contour_labeled.html#pyvista.ImageDataFilters.contour_labeled) to generate a surface from the binary image, and also smooth it (with `smoothing=True`)
5. finally, denoise the result even further with the [connectivity filter](https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.datasetfilters.connectivity#pyvista.DataSetFilters.connectivity), e.g extract largest
